### PR TITLE
chore: Upgrade to CUDA 12.1

### LINKED ISSUE #2010: Upgrade `main` stack to CUDA 12.1
With recent Torch nightly builds + TRT GA builds supporting CUDA 12.1, upgrade the stack to support new CUDA version

### PR DESCRIPTION
**Pull request by gs-olive**
# Description

- Upgrade CUDA version across CI, dependencies, and build systems
- Upgrade versions for Torch, TensorRT, and other dependencies in accordance with new CUDA version 12.1
- Upgrade Torch nightly date to latest available

Fixes #2010 

## Type of change

- Chore: version + dependency upgrades

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Comment by gs-olive** [comment]
Seems to be failing CI due to a strange interaction between `timm`, `torch`, and `tensorrt` imports. This is a minimal Segfaulting example:

```python
import tensorrt
import torch
import timm

model = (timm.create_model("efficientnet_b0", pretrained=True).eval().to("cuda"))
input_ = torch.randn((1, 3, 224, 224)).to("cuda")

# Segfaults, with the current CI machine configuration
model(input_)
```

**Comment by HolyWu** [comment]
> Seems to be failing CI due to a strange interaction between `timm`, `torch`, and `tensorrt` imports.

The segfault should be gone if you move `import tensorrt` below `import torch`, which means importing torch first and importing tensorrt later.

**Comment by narendasan** [comment]
> > Seems to be failing CI due to a strange interaction between `timm`, `torch`, and `tensorrt` imports.
> 
> The segfault should be gone if you move `import tensorrt` below `import torch`, which means importing torch first and importing tensorrt later.

Good catch but its probably not sufficient to just require folks to use a specific import order. @gs-olive what are the specific versions for tensorrt's canonical stack and pytorch's canonical stack?

**Comment by gs-olive** [comment]
Thanks for the catch @HolyWu. I am using `tensorrt==8.6.1` and `torch==2.1.0.dev20230613+cu121`. One notable difference in the requirements of the two packages is that the `tensorrt` Python package has `torch==1.11.0` in its requirements, [as shown here](https://github.com/NVIDIA/TensorRT/blob/a167852705d74bcc619d8fad0af4b9e4d84472fc/requirements.txt#L5-L9). None of PyTorch's required dependencies, [as shown here](https://github.com/pytorch/pytorch/blob/b1ddd5a2936ec796b0bb837709d83357083ff63a/requirements.txt), have strict requirements.

**Comment by gs-olive** [comment]
According to the dependency viewer, the `tensorrt` installation is not pulling in any linked `pip` dependencies. The `torch` installation, however, is using (in summary):

```python
torch                                  Tensors and Dynamic neural networks in Python with strong GPU acceleration
...
├── nvidia-cublas-cu11==11.10.3.66     CUBLAS native runtime libraries
│   ├── setuptools                     Easily download, build, install, upgrade, and uninstall Python packages
│   └── wheel                          A built-package format for Python
├── nvidia-cuda-cupti-cu11==11.7.101   CUDA profiling tools runtime libs.
│   ├── setuptools                     Easily download, build, install, upgrade, and uninstall Python packages
│   └── wheel                          A built-package format for Python
├── nvidia-cuda-nvrtc-cu11==11.7.99    NVRTC native runtime libraries
├── nvidia-cuda-runtime-cu11==11.7.99  CUDA Runtime native Libraries
│   ├── setuptools                     Easily download, build, install, upgrade, and uninstall Python packages
│   └── wheel                          A built-package format for Python
├── nvidia-cudnn-cu11==8.5.0.96        cuDNN runtime libraries
│   └── nvidia-cublas-cu11             CUBLAS native runtime libraries Metapackage
│       └── nvidia-cublas-cu117        CUBLAS native runtime libraries
│           ├── setuptools             Easily download, build, install, upgrade, and uninstall Python packages
│           └── wheel                  A built-package format for Python
├── nvidia-cufft-cu11==10.9.0.58       CUFFT native runtime libraries
├── nvidia-curand-cu11==10.2.10.91     CURAND native runtime libraries
│   ├── setuptools                     Easily download, build, install, upgrade, and uninstall Python packages
│   └── wheel                          A built-package format for Python
├── nvidia-cusolver-cu11==11.4.0.1     CUDA solver native runtime libraries
│   └── nvidia-cublas-cu11             CUBLAS native runtime libraries Metapackage
│       └── nvidia-cublas-cu117        CUBLAS native runtime libraries
│           ├── setuptools             Easily download, build, install, upgrade, and uninstall Python packages
│           └── wheel                  A built-package format for Python
├── nvidia-cusparse-cu11==11.7.4.91    CUSPARSE native runtime libraries
│   ├── setuptools                     Easily download, build, install, upgrade, and uninstall Python packages
│   └── wheel                          A built-package format for Python
├── nvidia-nccl-cu11==2.14.3           NVIDIA Collective Communication Library (NCCL) Runtime
├── nvidia-nvtx-cu11==11.7.91          NVIDIA Tools Extension
│   ├── setuptools                     Easily download, build, install, upgrade, and uninstall Python packages
│   └── wheel                          A built-package format for Python
...
```
The above seems strange because many of the dependencies are `cu11`, even though the torch distribution uses CUDA 12.

**Comment by gs-olive** [comment]
One other thing - the `torchvision` dependency (included in `timm`) is playing a role in the segmentation fault here. For example, the following does not fail:
```python
import tensorrt
import torch

class Model(torch.nn.Module):
    def forward(self, x):
        return x + 1

model = Model().eval().to("cuda")
input_ = torch.randn((1, 3, 224, 224)).to("cuda")
model(input_)
```

**Comment by HolyWu** [comment]
> One other thing - the `torchvision` dependency (included in `timm`) is playing a role in the segmentation fault here.

I think it has nothing to do with torchvision or timm. In the three Colab notebooks below, both torch 2.1.0.dev and torch 2.0.1 segfault. Only torch 2.0.0 doesn't fail.
https://colab.research.google.com/drive/1Oki2pnPT3GQbOzaDs0Jv0oeRTOMpBYLj?usp=sharing (torch 2.1.0.dev+cu121)
https://colab.research.google.com/drive/1ZLOFXe1ipfHIUz2XXqsnoMYQ_TUmiRdA?usp=sharing (torch 2.0.1+cu118)
https://colab.research.google.com/drive/19UxTgRwXfTyDe2qgFbtzB0M8Efzn0zCz?usp=sharing (torch 2.0.0+cu118)

File: toolchains/ci_workspaces/WORKSPACE.x86_64
Comment (gs-olive) [inline toolchains/ci_workspaces/WORKSPACE.x86_64:1]: Switching this `WORKSPACE` to be in line with all of the other `WORKSPACE` files, including those in `toolchains` and the main directory of the repository.

We need to pull `libtorch` directly from the `http` downloads path now because the distribution of Torch being installed through Python is lacking some necessary `c10` packages.

File: .circleci/config.yml [hunk 1]
```diff
@@ -186,21 +178,12 @@ commands:
...
-      cuda-pkg-version:
-        type: string
-        default: "11-8"
-      cuda-version:
-        type: string
-        default: "11.8.0"
       cuda-string-version:
         type: string
-        default: "cuda11.8"
+        default: "cuda12.0"
```
Comment (gs-olive) [inline .circleci/config.yml:183]: This version is `cuda12.0` and not `cuda12.1` like the others because the `apt` Ubuntu package directory does not have `nvinfer` libraries specifically for `cuda12.1`.

File: .circleci/config.yml [hunk 2]
```diff
@@ -261,27 +244,30 @@ commands:
...
+              pip3 install tensorrt==<< parameters.trt-version-long >> nvidia-cudnn-cu12==<< parameters.cudnn-version-long >>
               pip3 install pytest parameterized expecttest nox
 
   install-torch-from-index:
     description: "Install Pytorch"
     parameters:
       torch-build:
         type: string
-        default: "2.1.0.dev20230605+cu118"
+        default: "2.1.0.dev20230619+cu121.with.pypi.cudnn"
```
Comment (gs-olive) [inline .circleci/config.yml:255]: Using `.with.pypi.cudnn` resolves the issue detailed by @HolyWu - the sample notebooks provided here: https://github.com/pytorch/TensorRT/pull/2020#issuecomment-1601313192, were very helpful in debugging this issue.

Comment (HolyWu) [reply .circleci/config.yml:255]: Aye. The culprit is that torch 2.0.1 and 2.1.0.dev are having issue with `cuDNN 8.9.x`. When using `.with.pypi.cudnn` package, it installs `nvidia-cudnn-cu12 8.8.1.3` for you.

Comment (gs-olive) [reply .circleci/config.yml]: Thanks for the comment - this makes sense, and I've left the CuDNN version at `8.8`, which allows us to avoid usage of `.with.pypi.cudnn` here.

File: .circleci/config.yml [hunk 3]
```diff
@@ -92,17 +92,9 @@ commands:
...
-    steps:
-      - run:
-          name: "Uninstall CUDA-11.4"
-          command: |
-            sudo apt-get --purge remove "cuda*" "nsight*"
-            sudo apt-get --purge remove "*nvidia*"
 
   install-cudnn:
-    description: "Install CUDNN 8.8.0"
+    description: "Install CUDNN 8.8.1"
```
Comment (gs-olive) [inline .circleci/config.yml:97]: Based on feedback and discussion with @HolyWu and @narendasan, the CuDNN was only upgraded in the patch version to avoid issues with PyTorch and CuDNN 8.9, while preserving the `libtorch.so` version installed by Torch. A subsequent upgrade to the newest CuDNN is to come when PyTorch support is more robust.

**Review by narendasan — APPROVED** [review]
LGTM
